### ISSUE TITLE
Storable can't store VSTRINGs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12231]
Migrated from [rt.perl.org#113894](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113894) (status was 'resolved')

Searchable as RT113894$

**Comment by p5pRT** [comment]
### From @doy

#### Created by @doy



&nbsp;  use Storable 'dclone'
&nbsp;  my $copy = dclone\(\\v1\.2\.3\);
&nbsp;  is\(ref\($copy\)\, 'VSTRING'\);

This fails\, because Storable serializes the vstring as just a normal string\.
I've already added a TODO test for this in the Storable test suite\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    module=Storable

Site configuration information for perl 5.16.0:

Configured by doy at Sat Jun  2 13:46:21 CDT 2012.

Summary of my perl5 (revision 5 version 16 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=3.3.5-1-arch, archname=x86_64-linux
    uname='linux xtahua 3.3.5-1-arch #1 smp preempt mon may 7 19:57:51 cest 2012 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/doy/perl5/perlbrew/perls/perl-5.16.0'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.0 20120505 (prerelease)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.15.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.16.0:
    /home/doy/perl5/local/
    /home/doy/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0/x86_64-linux
    /home/doy/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0
    /home/doy/perl5/perlbrew/perls/perl-5.16.0/lib/5.16.0/x86_64-linux
    /home/doy/perl5/perlbrew/perls/perl-5.16.0/lib/5.16.0
    .


Environment for perl 5.16.0:
    HOME=/home/doy
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/doy/perl5/perlbrew/bin:/home/doy/perl5/perlbrew/perls/perl-5.16.0/bin:/home/doy/.bin/marathon:/home/doy/.bin/nethack:/home/doy/.bin/ghc:/home/doy/.bin:/usr/local/sbin:/usr/local/bin:/usr/lib/ccache/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/bin/vendor_perl:/usr/bin/core_perl
    PERL5LIB=/home/doy/perl5/local/
    PERLBREW_BASHRC_VERSION=0.33
    PERLBREW_HOME=/home/doy/.perlbrew
    PERLBREW_MANPATH=/home/doy/perl5/perlbrew/perls/perl-5.16.0/man
    PERLBREW_PATH=/home/doy/perl5/perlbrew/bin:/home/doy/perl5/perlbrew/perls/perl-5.16.0/bin
    PERLBREW_PERL=perl-5.16.0
    PERLBREW_ROOT=/home/doy/perl5/perlbrew
    PERLBREW_VERSION=0.43
    PERL_BADLANG (unset)
    PERL_CPANM_OPT=-q --mirror file:///home/doy/perl5/minicpan/ --mirror http://mirrors.kernel.org/cpan/ --mirror http://search.cpan.org/CPAN --prompt
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jun 28 20&#8203;:32&#8203;:02 2012\, doy@&#8203;tozt\.net wrote&#8203;:
>   use Storable 'dclone'
>   my $copy = dclone\(\\v1\.2\.3\);
>   is\(ref\($copy\)\, 'VSTRING'\);
> 
> This fails\, because Storable serializes the vstring as just a normal
>    string\.
> I've already added a TODO test for this in the Storable test suite\.

In case you are planning to fix this\, you might want to look at commit
d036e907fea\, which added vstring support to Data&#8203;::Dumper\.  Vstrings have
been a bit buggy in earlier versions of perl \(and Storable is
dual\-life\)\.  In particular\, ~v0 returns a scalar with vstring magic
attached to it when it shouldn’t\.  Data&#8203;::Dumper detects that and treats
it as not\-a\-vstring\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jun 28 21&#8203;:43&#8203;:30 2012\, sprout wrote&#8203;:
> On Thu Jun 28 20&#8203;:32&#8203;:02 2012\, doy@&#8203;tozt\.net wrote&#8203;:
> >   use Storable 'dclone'
> >   my $copy = dclone\(\\v1\.2\.3\);
> >   is\(ref\($copy\)\, 'VSTRING'\);
> > 
> > This fails\, because Storable serializes the vstring as just a normal
> >    string\.
> > I've already added a TODO test for this in the Storable test suite\.

I’ve fixed it with commits e00e3c3ee and 7e2a0d4586\.

> In case you are planning to fix this\, you might want to look at commit
> d036e907fea\, which added vstring support to Data&#8203;::Dumper\.  Vstrings have
> been a bit buggy in earlier versions of perl \(and Storable is
> dual\-life\)\.  In particular\, ~v0 returns a scalar with vstring magic
> attached to it when it shouldn’t\.  Data&#8203;::Dumper detects that and treats
> it as not\-a\-vstring\.

That did not actually matter\, as Storable can store both the PV and the
magic PV\, so what comes back is identical\.

Data&#8203;::Dumper can only store one \(either "\\1\\2\\3" or v1\.2\.3\)\, so it has
to choose based on whether they actually match\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
